### PR TITLE
[FEATURE]  "Duplicate feature redigitized" in MapLayerActions

### PR DESCRIPTION
**Pull request by signedav**
## Description
Functionality to duplicate features with the original geometry and with digitizing new geometry. And still duplicate selected features in attribute table.

1. "Duplicate feature" is now a MapLayerAction.
The partwise obsolete Python Default Action "Duplicate feature" is still available on feature scope (attributetable), because there it's not obsolete at all. It's called now "Duplicate selected features" - it's needed to duplicate all the selected features. The MapLayerAction is only able to duplicate one feature, because with redigitizing it would be unhandy for users if there are more than one feature.
duplicateselected

2. The functionality "Duplicate feature redigitized"
Means we can now duplicate a feature with digitizing the geometry of the duplicate.
duplicategeom

There are new icons for these actions - I hope they are okay.

### CODE REVIEW

**Comment by nyalldawson** [comment]
I think we need some better wording for "duplicate feature redigitized"... but I can't think of any offhand :P

**Comment by 3nids** [comment]
duplicates attributes and digitize?

**Comment by signedav** [comment]
I don't mind better wording.

"duplicate attribute and digitize" is a possibility. It's actually not exactly what it does, but for a user there is no difference, whether the feature is duplicated and then digitized or an instance is digitized first and then duplicated (with new geometry). My concern here is more, if it's too long.

Btw. why use the word attribute instead of feature?

Actually we have three types of duplication:
- Duplicate feature
- Duplicate feature redigitized
- Duplicate selected features (only in the attribute table - layer scope)

So "duplicate attribute", "duplicate attribute and digitize", "duplicate selected attributes" would be okay for me. But probably there are better ones.

**Comment by m-kuhn** [comment]
Other proposals:

* Duplicate feature with new geometry
* Duplicate feature and redraw geometry
* Duplicate feature and digitize a geometry with a NEW (in 3.2!!) map tool. Earns bitcoins for the QGIS project if used appropriately

I'd stick to the term "feature"

**Comment by signedav** [comment]
I'm fine with "and digitize" from @3nids or the "and redraw geometry" from @m-kuhn if it's not too long in your opinion. 
But I prefer "feature" to "attribute".

**Comment by signedav** [comment]
My proposal:

**Comment by signedav** [comment]
I decided to make it like this
- Duplicate feature
- Duplicate feature and digitize
- Duplicate selected features (only in the attribute table - layer scope)

**Comment by andreasneumann** [comment]
sounds good. Good choice!

File: src/app/qgisapp.cpp [hunk 1]
```diff
@@ -13301,3 +13328,104 @@ void QgisApp::transactionGroupCommitError( const QString &error )
...
 }
+
+QgsFeature QgisApp::duplicateFeatures( QgsMapLayer *mlayer, const QgsFeature &feature )
+{
+  if ( mlayer->type() != QgsMapLayer::VectorLayer )
+    return QgsFeature();
+
+  QgsVectorLayer *layer = qobject_cast<QgsVectorLayer *>( mlayer );
+
+  layer->startEditing();
```
Comment (pblottiere) [inline src/app/qgisapp.cpp]: What do you think about checking that `layer` is not null in this case (just after casting `mLayer` in a vector layer)?

Comment (pblottiere) [reply src/app/qgisapp.cpp]: Sorry, I didn't see that you previously checked the type of layer...

**Comment by pblottiere** [comment]
Hi @signedav, nice feature!

**Comment by nyalldawson** [comment]
@signedav @m-kuhn just hit an issue with this - the action is enabled in the UI even when a layer is non editable. It's even enabled for layers which should be read-only. (same with the "Duplicate feature" action)

Filed as https://issues.qgis.org/issues/17852

**Comment by signedav** [comment]
Thanks @nyalldawson 
This is on purpose. I thought it should behave like the user actions. These are always in the UI. 
The duplicate-actions enable the editable mode, when they are selected. So this makes kind of sense.

But if this is not allowed, it could be changed. There is the possibility to make them active in the UI depending on editable mode or to raise a message when it's not in editable mode like: "Duplicate and digitize feature can only be done for layers in editable mode."

On read-only layers I agree it should be blocked anyway. Does not matter for what solution.

**Comment by nyalldawson** [comment]
@signedav

I'd say they should work like other edit tools instead. Skipping the edit buffer is really quite dangerous here. 

My preference would be to only enable these actions in edit mode, in order to keep consistency with all the other edit tools

**Comment by m-kuhn** [comment]
> Skipping the edit buffer

I think they toggle the edit mode.

**Comment by nyalldawson** [comment]
> I think they toggle the edit mode.

It didn't for me. And when testing with a gdb file (which can't be made editable) it happily reported that the feature was duplicated!

**Comment by signedav** [comment]
Okay. I can care about that.

**Pull request merged**